### DISCUSSION TITLE
[Help] List notes with broken links

### DISCUSSION THREAD

**Discussion by kirasok** [zk-org/zk#335] [Help] [answered]
### Verify

- [X] I searched the existing discussions for help

### How can we help you?

Hi! Recently I renamed some of my notes. It broke links to them in other notes. Is there way to list notes that have broken links in them?

**Comment by mickael-menu** [comment]
Maybe there's a way by parsing the output of `zk graph --format json`, but nothing to do that out of the box.

**Comment by prikhi** [comment]
I played around a little with jq + regular expressions(to match `[[Title]]` or `[[Title|other text]]` formats), used that to build a list of all link titles & then did the list difference between all link titles & all titles:

```sh
$ zk graph --format json 2>/dev/null | jq '{ allLinks: .notes|map([.body|scan("\\[\\[(.*?)(\\|.*?)?]]")|.[0]])|add|unique, allTitles: .notes|map(.title) }| .allLinks - .allTitles'
```

**Reply by kirasok** [reply, marked as answer]
Thank you! I've changed `allTitles: .notes|map(.title)` to `allTitles: .notes|map(.filenameStem)` so it would work in cases when title and filename don't match

**Solution** to print missing notes:
```bash
zk graph --format json 2>/dev/null | jq '{ allLinks: .notes|map([.body|scan("\\[\\[(.*?)(\\|.*?)?]]")|.[0]])|add|unique, allTitles: .notes|map(.filenameStem) }| .allLinks - .allTitles'
```

and zk alias:
```yaml
[alias]
broken = "zk graph --format json 2>/dev/null | jq '{ allLinks: .notes|map([.body|scan(\"\\\\[\\\\[(.*?)(\\\\|.*?)?]]\")|.[0]])|add|unique, allTitles: .notes|map(.filenameStem) }| .allLinks - .allTitles'"
```

**Reply by gabevenberg** [reply]
As someone who is really not good with jq, how would you go about listing the notes that contain a broken link?
